src/commands/protocols/common/handle-protocol-message-command.js

```diff
@@ -40,6 +45,161 @@ class HandleProtocolMessageCommand extends Command {
+                tokenId,
+                keyword,
+                hashFunctionId,
+            );
+            this.logger.info(
+                `Calculated agreement id: ${agreementId} for contract: ${contract}, token id: ${tokenId}, keyword: ${keyword}, hash function id: ${hashFunctionId}, operationId: ${operationId}`,
+            );
+
+            return {
+                agreementId,
+                agreementData: await this.blockchainModuleManager.getAgreementData(
+                    blockchain,
+                    agreementId,
+                ),
+            };
+        };
+
+        const getAsk = async () => {
+            const peerRecord = await this.repositoryModuleManager.getPeerRecord(
+                this.networkModuleManager.getPeerId().toB58String(),
+                blockchain,
+            );
+
+            const ask = this.blockchainModuleManager.convertToWei(blockchain, peerRecord.ask);
+
+            return this.blockchainModuleManager.toBigNumber(blockchain, ask);
+        };
+
+        const [{ agreementId, agreementData }, blockchainAssertionSize, r0, ask] =
+            await Promise.all([
+                geAgreementData(),
+                this.blockchainModuleManager.getAssertionSize(blockchain, assertionId),
+                this.blockchainModuleManager.getR0(blockchain),
+                getAsk(),
+            ]);
+
+        const divisor = this.blockchainModuleManager
+            .toBigNumber(blockchain, r0)
+            .mul(agreementData.epochsNumber)
+            .mul(blockchainAssertionSize);
+
+        const serviceAgreementBid = this.blockchainModuleManager
+            .toBigNumber(blockchain, agreementData.tokenAmount)
+            .add(agreementData.updateTokenAmount)
+            .mul(1024)
+            .div(divisor)
+            .add(1); // add 1 wei because of the precision loss
+
+        const bidAskLog = `Service agreement bid: ${serviceAgreementBid}, ask: ${ask}, operationId: ${operationId}`;
+        this.logger.trace(bidAskLog);
+
+        return {
+            errorMessage: ask.lte(serviceAgreementBid) ? null : bidAskLog,
+            agreementId,
+            agreementData,
+        };
+    }
+
+    async validateReceivedData(
+        operationId,
+        assertionId,
+        blockchain,
+        contract,
+        tokenId,
+        keyword,
+        hashFunctionId,
+    ) {
+        const ual = this.ualService.deriveUAL(blockchain, contract, tokenId);
+
+        this.logger.trace(`Validating neighborhood for ual: ${ual}`);
+        if (!(await this.validateNeighborhood(blockchain, keyword, hashFunctionId, ual))) {
+            return {
+                messageType: NETWORK_MESSAGE_TYPES.RESPONSES.NACK,
+                messageData: { errorMessage: 'Invalid neighbourhood' },
+            };
+        }
+
+        this.logger.trace(`Validating assertion with ual: ${ual}`);
+        await this.validateAssertionId(blockchain, contract, tokenId, assertionId, ual);
```

Comment:
This won't work for update? Because assertionId isn't pushed to array until state isn't finalized

Reply:
I see that we're overwriting it for update, but it makes no sense to me, can we make it universal?

Reply:
validation is done differently for update and regular assertion, it's normal to use method overwriting for that case